qiskit/transpiler/passes/synthesis/solovay_kitaev.py

```diff
@@ -0,0 +1,433 @@
+        self._sk = SolovayKitaev(basis_gates, depth)
+
+    @staticmethod
+    def _get_default_approximations():
+        from .depth10_t_h_tdg import decompositions
+
+        return decompositions
+
+    @staticmethod
+    def generate_basic_approximations(
+        basis_gates: List[Union[str, Gate]], depth: int, filename: str
+    ) -> None:
+        """Generate and save the basic approximations.
+
+        Args:
+            basis_gates: The gates from which to create the sequences of gates.
+            depth: The maximum depth of the approximations.
+            filename: The file to store the approximations.
+        """
+        SolovayKitaev.generate_basic_approximations(basis_gates, depth, filename)
+
+    def load_basic_approximations(self, filename: str) -> None:
+        """Load basic approximations.
+
+        Args:
+            filename: The path to the file from where to load the data.
+        """
+        self._sk.load_basic_approximations(filename)
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the SolovayKitaevDecomposition pass on `dag`.
+
+        Args:
+            dag: The input dag.
+
+        Returns:
+            Output dag with 1q gates synthesized in the discrete target basis.
+
+        Raises:
+            TranspilerError: if a gates does not have to_matrix
+        """
+        if self._sk._basic_approximations is None:
+            self.load_basic_approximations(self._get_default_approximations())
+
+        for node in dag.op_nodes():
+            if not node.op.num_qubits == 1:
+                continue  # ignore all non-single qubit gates
+
+            if not hasattr(node.op, "to_matrix"):
+                raise TranspilerError(
+                    "SolovayKitaevDecomposition does not support gate without "
+                    f"to_matrix method: {node.op.name}"
+                )
+
+            matrix = node.op.to_matrix()
+
+            # call solovay kitaev
+            approximation = self._sk.run(matrix, self.recursion_degree)
+
+            # convert to a dag and replace the gate by the approximation
+            substitute = circuit_to_dag(approximation)
+            dag.substitute_node_with_dag(node, substitute)
+
+        return dag
+
+
+def _check_candidate(candidate: GateSequence, sequences: List[GateSequence]) -> bool:
+    from qiskit.quantum_info.operators.predicates import matrix_equal
+
+    # check if a matrix representation already exists
+    for existing in sequences:
+        # eliminate global phase
+        if matrix_equal(existing.product, candidate.product, ignore_phase=True):
+            # is the new sequence less or more efficient?
+            if len(candidate.gates) >= len(existing.gates):
+                return False
+            return True
+    return True
+
+
+def _remove_inverse_follows_gate(sequence):
+    index = 0
+    while index < len(sequence.gates) - 1:
+        if sequence.gates[index + 1] == sequence.gates[index].inverse():
```

Comment:
This could potentially be slow. Maybe we can improve this by not checking the gates themselves but rather their names or types

Reply:
Yes indeed, that improved the runtime by a factor of a bit more than 2.